samza-core/src/main/java/org/apache/samza/table/remote/RemoteReadWriteTable.java

```diff
@@ -69,33 +67,36 @@ public void init(SamzaContainerContext containerContext, TaskContext taskContext
     super.init(containerContext, taskContext);
     writeMetrics = new DefaultTableWriteMetrics(containerContext, taskContext, this, tableId);
     TableMetricsUtil tableMetricsUtil = new TableMetricsUtil(containerContext, taskContext, this, tableId);
-    putThrottleNs = tableMetricsUtil.newTimer("put-throttle-ns");
+    writeRateLimiter.setTimerMetric(tableMetricsUtil.newTimer("put-throttle-ns"));
   }
 
   /**
    * {@inheritDoc}
    */
   @Override
   public void put(K key, V value) {
+    try {
+      putAsync(key, value).get();
+    } catch (Exception e) {
+      throw new SamzaException(e);
+    }
+  }
 
+  /**
+   * {@inheritDoc}
+   */
+  @Override
+  public CompletableFuture<Void> putAsync(K key, V value) {
+    Preconditions.checkNotNull(key);
     if (value == null) {
-      delete(key);
-      return;
+      return deleteAsync(key);
     }
 
-    try {
-      writeMetrics.numPuts.inc();
-      if (rateLimitWrites) {
-        throttle(key, value, RL_WRITE_TAG, writeCreditFn, putThrottleNs);
-      }
-      long startNs = System.nanoTime();
-      writeFn.put(key, value);
-      writeMetrics.putNs.update(System.nanoTime() - startNs);
-    } catch (Exception e) {
-      String errMsg = String.format("Failed to put a record, key=%s, value=%s", key, value);
-      logger.error(errMsg, e);
-      throw new SamzaException(errMsg, e);
-    }
+    writeMetrics.numPuts.inc();
+    return execute(writeRateLimiter, key, value, writeFn::putAsync, writeMetrics.putNs)
```

Comment:
maybe: "execute" seems a bit general for this function name. Maybe name it doPutAsync or something to that effect.

Some of these are class variables right? So, you should be able to simplify some Function pointers to that execute call.